csp_billing_adapter_microsoft/plugin.py

```diff
@@ -71,6 +70,102 @@ def meter_billing(
     is attempted 3 times before re-raising the exception to
     calling scope.
     """
+    metered_billing_data = {"request": []}
+    status = {}
+
+    for dimension_name, quantity in dimensions.items():
+        if quantity == 0:
+            log.info(
+                'A "0" value was reported for %s, skipping meter billing',
+                dimension_name
+            )
+            continue
+
+        # Setup request body for the usage event API
+        metered_billing_data["request"].append(
+            {
+                'resourceUri': os.environ['EXTENSION_RESOURCE_ID'],
+                'quantity': quantity,
+                'dimension': dimension_name,
+                'effectiveStartTime': str(timestamp.astimezone(timezone.utc)),
+                'planId': os.environ['PLAN_ID']
+            }
+        )
+
+    if metered_billing_data["request"]:
+        # Setup request header for the usage event API
+        metered_billing_header = {
+            'Content-type': 'application/json',
+            'x-ms-correlationid': str(uuid.uuid4()),
+            'authorization': _get_msi_token()
+        }
+
+        # Make request to the usage event API
+        data_request = urllib.request.Request(
+            'https://marketplaceapi.microsoft.com/api/batchUsageEvent'
+            '?api-version=2018-08-31',
+            data=json.dumps(metered_billing_data).encode(),
+            headers=metered_billing_header,
+            method='POST'
+        )
+
+        retries = 3
+        exc = None
+        response = None
+
+        while retries > 0:
+            try:
+                with urllib.request.urlopen(data_request) as url_open_return:
+                    response = json.loads(
+                        url_open_return.read().decode("utf-8")
+                    )
+                    break
+            except urllib.error.URLError as error:
+                exc = error
+                retries -= 1
+                continue
+
+        if exc:
+            msg = (
+                f"Failed to meter bill dimensions "
+                f"{dimensions}: {str(exc)}"
+            )
+            for dimension_name in dimensions:
+                status[dimension_name] = {
+                    "status": "failed",
+                    "error": msg
+                }
+            log.error(msg)
+            return status
+
+        if response and (response.get("count", 0) > 0):
+            for resp in response["result"]:
+                if resp["status"] == "Accepted":
```

Comment:
Minor readability suggestion; you could simplify this code somewhat by having the if/else branches just construct the dictionary variable, e.g. dim_status, holding the status and either record_id or error entry, and then assign that dictionary variable to status[resp["dimension"]] immediately after the else block

Reply:
Thanks, I will clean this up

Reply:
please re-review

Reply:
Looks good